File: Python/Cumpleaños/Cumpleaños.py
from datetime import datetime

# Función para calcular la edad y los días restantes para el próximo cumpleaños
def calcular_edad_y_dias_restantes(fecha_nacimiento):
    # Obtener la fecha actual
    fecha_actual = datetime.now()
    
    # Convertir la fecha de nacimiento a un objeto datetime
    fecha_nacimiento = datetime.strptime(fecha_nacimiento, "%d-%m-%Y")
    
    # Calcular la diferencia de años
    diferencia_anios = fecha_actual.year - fecha_nacimiento.year
    
    # Comprobar si ya pasó el cumpleaños de este año
    if (fecha_actual.month, fecha_actual.day) < (fecha_nacimiento.month, fecha_nacimiento.day):
        diferencia_anios -= 1
    
    # Calcular la fecha del próximo cumpleaños
    proximo_cumple = datetime(fecha_actual.year, fecha_nacimiento.month, fecha_nacimiento.day)
    
    # Calcular los días restantes para el próximo cumpleaños
    dias_restantes = (proximo_cumple - fecha_actual).days
    
    return diferencia_anios, dias_restantes

# Pedir al usuario que ingrese la fecha de nacimiento
fecha_nacimiento = input("Ingresa tu fecha de nacimiento (dd-mm-yyyy): ")

# Calcular la edad y los días restantes
edad, dias_restantes = calcular_edad_y_dias_restantes(fecha_nacimiento)

# Obtener la fecha actual
fecha_actual = datetime.now()

# Comprobar si el cumpleaños es hoy
es_cumple_hoy = (fecha_actual.strftime("%d-%m-%Y") == fecha_nacimiento)

# Mostrar la edad y los días restantes
print(f"Tienes {edad} años.")
if es_cumple_hoy:
    print("¡Feliz cumpleaños!")
else:
    print(f"Faltan {dias_restantes} días para tu próximo cumpleaños.")

File: Java/Cumpleaños/src/Cumpleaños.java
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Scanner;

public class Cumpleaños {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Ingresa tu fecha de nacimiento (dd-mm-yyyy): ");
        String fechaNacimientoStr = scanner.nextLine();

        SimpleDateFormat dateFormat = new SimpleDateFormat("dd-MM-yyyy");
        Date fechaNacimiento;

        try {
            fechaNacimiento = dateFormat.parse(fechaNacimientoStr);
            Calendar calendarNacimiento = Calendar.getInstance();
            calendarNacimiento.setTime(fechaNacimiento);

            // Obtener la fecha actual
            Calendar fechaActual = Calendar.getInstance();

            // Calcular la edad restando la fecha de nacimiento a la fecha actual
            int edad = fechaActual.get(Calendar.YEAR) - calendarNacimiento.get(Calendar.YEAR);

            // Verificar si ya ha pasado el cumpleaños de este año
            if (fechaActual.get(Calendar.MONTH) < calendarNacimiento.get(Calendar.MONTH)
                    || (fechaActual.get(Calendar.MONTH) == calendarNacimiento.get(Calendar.MONTH)
                    && fechaActual.get(Calendar.DAY_OF_MONTH) < calendarNacimiento.get(Calendar.DAY_OF_MONTH))) {
                edad--;
            }

            // Calcular la fecha del próximo cumpleaños
            Calendar proximoCumpleanos = (Calendar) calendarNacimiento.clone();
            proximoCumpleanos.add(Calendar.YEAR, edad + 1);

            // Calcular la cantidad de días hasta el próximo cumpleaños
            long tiempoRestanteMillis = proximoCumpleanos.getTimeInMillis() - fechaActual.getTimeInMillis();
            int diasFaltantes = (int) (tiempoRestanteMillis / (1000 * 60 * 60 * 24));

            // Verificar si hoy es el cumpleaños
            boolean esCumpleanosHoy = fechaActual.get(Calendar.MONTH) == calendarNacimiento.get(Calendar.MONTH)
                    && fechaActual.get(Calendar.DAY_OF_MONTH) == calendarNacimiento.get(Calendar.DAY_OF_MONTH);

            // Mostrar la edad y días hasta el próximo cumpleaños
            System.out.println("Tienes " + edad + " años.");
            System.out.println("Faltan " + diasFaltantes + " días para tu próximo cumpleaños.");

            // Mostrar mensaje de felicitación si es el cumpleaños
            if (esCumpleanosHoy) {
                System.out.println("¡Feliz cumpleaños!");
            }
        } catch (ParseException e) {
            System.out.println("Fecha de nacimiento no válida. Por favor, ingresa una fecha en formato válido (dd-MM-yyyy).");
        }
    }
}
